doc/api/esm.md

```diff
@@ -191,6 +191,14 @@ also be used within `"exports"` to define different package entry points per
 environment, including whether the package is referenced via `require` or via
 `import`.
 
+**Warning**: Introducing the `"exports"` field prevents consumers of a package
+from using any entry points that are not defined, including the `package.json`.
+This will likely be a breaking change. To make the introduction of `"exports"`
+non-breaking, ensure that every previously supported entry point is exported.
+It is best to explicitly specify entry points so that the package's public API is
+well-defined. As a last resort, package encapsulation can be disabled entirely by
```

Comment:
Could we perhaps include an explicit example here? Something like:

> For example, a CommonJS package with `lib/index.js` and `feature/index.js` where users might have previously 
imported either subpath explicitly without explicitly including the `index.js` path, can be written:

```
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

> Since all common existing require patterns are captured in the above, it can be a non-breaking change.

> If you must adopt `"exports"` and really cannot explicitly list subpaths or make a breaking change, then a `"./": "./"` root export can be provided to opt-out of encapsulation.

Perhaps it is overkill, but I just worry a wall of text with one code example like `"./": "./"` will lead to users just copying that directly as the "I just want this to work" path when scanning, unless they can explicitly see what this really involves.